src/org/rust/lang/core/resolve/RustResolveEngine.kt

```diff
@@ -68,6 +79,26 @@ public class RustResolveEngine() {
             seek(scope.getDeclarations())
         }
 
+        private fun processUseDeclaration(use: RustUseItem) {
+            val path = use.viewPath
+            val pathPart = path.pathPart ?: return
+
+            val isPlainPathImport = path.mul == null && path.lbrace == null
+            if (isPlainPathImport) {
+                val name = path.rename ?: pathPart
+                if (match(name)) {
+                    if (use in visited) {
+                        return
+                    }
+                    visited += use
```

Comment:
That is, we can check `if (use in visited)` before match, but it is crucial that we do `visited += use` after match.